File: Leetcode Daily Challenge/016-October-2024/014-maximal-score-after-applying-k-operations.java
// https://leetcode.com/problems/maximal-score-after-applying-k-operations/


import java.util.*;

class Solution {
    public long maxKelements(int[] nums, int k) {
        long ans = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());

        for (int num : nums) {
            pq.offer(num);
        }

        while (k -- > 0) {
            int topEl = pq.poll();
            ans += topEl;
            
            pq.offer( (topEl + 2) / 3 );
        }
        return ans;
    }
}

File: Leetcode Daily Challenge/016-October-2024/014-maximal-score-after-applying-k-operations.cpp
// https://leetcode.com/problems/maximal-score-after-applying-k-operations/



#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TC : O(N + K*log(N))
    // SC : O(N)
    long long maxKelements(vector<int>& nums, int k) {
        long long ans = 0;
        priority_queue<int> pq(begin(nums), end(nums)); // O(N) only in cpp heapify

        while (k --) {
            int topEl = pq.top();
            pq.pop();

            ans += topEl;

            pq.push( (topEl + 2) / 3 );
        }
        return ans;
    }
};